pages/instructorAssessmentManualGrading/instructorAssessmentManualGrading.ejs

```diff
@@ -51,10 +56,29 @@
                 <td class="text-center">
                   <%= question.num_submissions %>
                 </td>
+                <td class="text-center">
+                  <%= question.num_submissions_eligible %>
+                </td>
+                <td class="text-center">
+                  <%= question.num_graded %>
+                </td>
+                <td class="text-center">
+                  <%= question.num_locked_submissions %>
+                </td>
+                <td class="text-center">
+                  <%= question.manual_grading_users %>
+                </td>
+                <td class="text-center">
+                  <% if (question.num_locked_submissions > 0) { %>
+                    <a class="btn btn-primary" href="<%= urlPrefix %>/assessment/<%= assessment.id %>/assessment_question/<%= question.id %>/unlock_ungraded">
+                      Unlock <%= question.num_locked_submissions %> submission<% if (question.num_locked_submissions > 1) { %>s<% } %>
```

Comment:
If we restrict each user to only having one question at a time locked for grading then we could avoid all this unlocking code.

Reply:
We could eliminate this code for sure, but it might come back in another way. For example, suppose we were to lock one item:

1) What happens if a user closes the browser? The item would still be locked.
  a) User is redirected back to question to finish grading on starting manual grading again (if and only if they start manual grading again. Hence, an item could be locked and forgotten, which requires further logic)
    i) Other users take over locked items after x period of time set in config
    ii) Other users take over locked items if there are no more items to be graded --> Could possibly run into a grading conflict here.

It would be ideal if we could lock an item at the database level and let it manage the lock-- eventually releasing it. The for update lock will end after the transaction ends, however. 

Please let me know if you are not concerned with edge cases here, as we can try it out and see if it works without this code.

Reply:
DRAFT: I'm going to try to get this right, and we'll see if it's helpful. **Note**, each TA can have only one question instance open for grading, and that is the one most recently opened. Every other question instance a TA has opened should be considered stale. Any question instance which is not _currently_ open for grading, including those which are stale, _can_ be open for grading, and thus selected in the "to be graded" list. 

Suppose there are 2 TAs, and TA1 has a qinst open for grading. If TA2 opens the same qinst (somehow) they do so under a warning, but nothing changes in their access ... they can grade and even submit the grade, under the warning. When any TA submits a score, if there has been a state change in the score since the qinst was open for grading, then the TA is offered a choice between the two? scores, and a new submission is created if they insist on their own score. Possibilities for race conditions 1) two TAs insist on their own score (and thus make a new submission) -- this is a human problem and bears conversation. Do we want to facilitate that in any way? 2) both TAs defer and thus no score submission is made, at all. IMO this should just put the qinst back into the list of things to be graded. Is this always a pairwise problem? that is, can we always assume that the choice is between the current grade and the most recently submitted one?

Reply:
> DRAFT: I'm going to try to get this right, and we'll see if it's helpful. **Note**, each TA can have only one question instance open for grading, and that is the one most recently opened. Every other question instance a TA has opened should be considered stale. Any question instance which is not _currently_ open for grading, including those which are stale, _can_ be open for grading, and thus selected in the "to be graded" list.
> 
> Suppose there are 2 TAs, and TA1 has a qinst open for grading. If TA2 opens the same qinst (somehow) they do so under a warning, but nothing changes in their access ... they can grade and even submit the grade, under the warning. When any TA submits a score, if there has been a state change in the score since they opened the qinst for grading, then the TA is offered a choice between the two? scores, and a new submission is created if they insist on their own score. Possibilities for race conditions 1) two TAs insist on their own score (and thus make a new submission) -- this is a human problem and bears conversation. Do we want to facilitate that in any way? 2) both TAs defer and thus no score submission is made, at all. IMO this should just put the qinst back into the list of things to be graded. Is this always a pairwise problem? that is, can we always assume that the choice is between the current grade and the most recently submitted one?

Regarding your first paragraph, we need to then:

1) Any time a user opens an instance question for manual grading, we can: 

 - remove grading user from any instance questions that have been opened for manual grading by this user and have been ungraded/abandoned
 
But we cannot tell when an instance question is currently open and being graded without adding more logic to send requests to the server on an interval or abandoning a question after 'x' amount of time. Hence, my last comment to Matt on this thread.

2) I would propose that we send a request to the back-end and if the submission has been graded, we can create a modal with the diff grading info. This would require more code though, so @mwest1066 would have to approve, as he is currently looking for less code for this feature.

I think the general idea would be to hope that TAs communicate if they are both grading the same question so they do not run into a conflict in grading. If they both submit a grade, the latter submission overwrites the former submission without any warnings. Perhaps that would be a good warning banner for another user grading the same question?